<a href="https://colab.research.google.com/github/MatheusSteinDeAguiar/Regex_MLE_Laplace/blob/main/Introducao_a_regex_e_modelos_de_linguagem.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Inicializando

In [1]:
!pip install nltk==3.5

In [2]:
import pandas as pd
import re

###Carregando os dados

In [3]:
dados_pt = pd.read_csv("stackoverflow_portugues.csv")
dados_ing = pd.read_csv("stackoverflow_ingles.csv")
dados_sp = pd.read_csv("stackoverflow_espanhol.csv")

In [4]:
dados_pt.head()

Id  ... Visualizações
0   2402  ...         22367
1   6441  ...        176953
2    579  ...          9761
3   2539  ...          5075
4  17501  ...         54191

[5 rows x 6 columns]

In [5]:
questao_1_pt = dados_pt.Questão[0]
questao_1_ing = dados_ing.Questão[0]
questao_1_sp = dados_sp.Questão[0]

print(questao_1_pt)

<p>Se eu fizer o <em><a href="http://pt.wikipedia.org/wiki/Fun%C3%A7%C3%A3o_de_embaralhamento_criptogr%C3%A1fico" rel="noreferrer">hash</a></em> de senhas antes de armazená-las em meu banco de dados é suficiente para evitar que elas sejam recuperadas por alguém?</p>

<p>Estou falando apenas da recuperação diretamente do banco de dados e não qualquer outro tipo de ataque, como <a href="http://pt.wikipedia.org/wiki/Ataque_de_for%C3%A7a_bruta" rel="noreferrer">força bruta</a> na página de login da aplicação, <em><a href="http://pt.wikipedia.org/wiki/Keylogger" rel="noreferrer">keylogger</a></em> no cliente e <a href="http://pt.wikipedia.org/wiki/Criptoan%C3%A1lise_de_mangueira_de_borracha" rel="noreferrer">criptoanálise <em>rubberhose</em></a>. Qualquer forma de <em>hash</em> não vai impedir esses ataques.</p>

<p>Tenho preocupação em dificultar ou até impossibilitar a obtenção das senhas originais caso o banco de dados seja comprometido. Como dar maior garantia de segurança neste aspecto

#Regular Expressions

In [6]:
#Regex para remover todas as tags html
re.findall(r"<.*?>", questao_1_pt)

['<p>',
 '<em>',
 '<a href="http://pt.wikipedia.org/wiki/Fun%C3%A7%C3%A3o_de_embaralhamento_criptogr%C3%A1fico" rel="noreferrer">',
 '</a>',
 '</em>',
 '</p>',
 '<p>',
 '<a href="http://pt.wikipedia.org/wiki/Ataque_de_for%C3%A7a_bruta" rel="noreferrer">',
 '</a>',
 '<em>',
 '<a href="http://pt.wikipedia.org/wiki/Keylogger" rel="noreferrer">',
 '</a>',
 '</em>',
 '<a href="http://pt.wikipedia.org/wiki/Criptoan%C3%A1lise_de_mangueira_de_borracha" rel="noreferrer">',
 '<em>',
 '</em>',
 '</a>',
 '<em>',
 '</em>',
 '</p>',
 '<p>',
 '</p>',
 '<p>',
 '<em>',
 '</em>',
 '</p>']

In [7]:
texto_teste = re.sub(r"<.*?>", "", questao_1_pt)
print(texto_teste)

Se eu fizer o hash de senhas antes de armazená-las em meu banco de dados é suficiente para evitar que elas sejam recuperadas por alguém?

Estou falando apenas da recuperação diretamente do banco de dados e não qualquer outro tipo de ataque, como força bruta na página de login da aplicação, keylogger no cliente e criptoanálise rubberhose. Qualquer forma de hash não vai impedir esses ataques.

Tenho preocupação em dificultar ou até impossibilitar a obtenção das senhas originais caso o banco de dados seja comprometido. Como dar maior garantia de segurança neste aspecto?

Quais preocupações adicionais evitariam o acesso às senhas? Existem formas melhores de fazer esse hash?



In [8]:
# Se compilarmos a regex antes e usar ela para aplicar, podemos diminuir muito o tempo de execução.
regex = re.compile(r"<.*?>")
texto_teste = regex.sub("", questao_1_pt)
print(texto_teste)

Se eu fizer o hash de senhas antes de armazená-las em meu banco de dados é suficiente para evitar que elas sejam recuperadas por alguém?

Estou falando apenas da recuperação diretamente do banco de dados e não qualquer outro tipo de ataque, como força bruta na página de login da aplicação, keylogger no cliente e criptoanálise rubberhose. Qualquer forma de hash não vai impedir esses ataques.

Tenho preocupação em dificultar ou até impossibilitar a obtenção das senhas originais caso o banco de dados seja comprometido. Como dar maior garantia de segurança neste aspecto?

Quais preocupações adicionais evitariam o acesso às senhas? Existem formas melhores de fazer esse hash?



In [9]:
# Definindo uma função capaz de substiuir o conteúdo da regex por espaço vazio seja 1 ou mais textos.
def remover(regex, textos):
  if type(textos) == str:
    return regex.sub("", textos)
  else:
    return [regex.sub("", texto) for texto in textos]

In [10]:
# A regex escrita usa os sinais <> porque as tags estão sempre dentro deles, a combinação de .* porque o ponto é 
# qualquer caractere e o * é quantas vezes for necessário, e o ? para que pare quando encontrar o ">", se não 
# acabaria captando o texto junto.
regex_html = re.compile(r"<.*?>")
# Antes de remover as tags html é interessante remover o código visto que através das tags 
# ele é facilmente identificado.
# Não podemos deixar a regex apenas com o .* pois as quebras de linha fariam a regex parar, então precisamos colocar
# que identifique qualquer caractere ou quebra de linha (\n) entre o inicio e o fim do código.
regex_cod = re.compile(r"<code>(.|(\n))*?</code>")

In [11]:
def substitui(regex, textos, substituto):
    if type(textos) == str:
        return regex.sub(substituto, textos)
    else:
        return [regex.sub(substituto, texto) for texto in textos]

Tirando o código e as tags html's das questões em português.

In [12]:
questao_1_sp_sem_code = substitui(regex_cod, questao_1_sp, "CODE")
print(questao_1_sp_sem_code)

<p>Las sentencias dinámicas son sentencias SQL que se crean como cadenas de texto (strings) y en las que se insertan/concatenan valores obtenidos de alguna fuente (normalmente proveniente del usuario), lo que puede hacer que sean vulnerables a inyección SQL si no se sanean las entradas, como por ejemplo:</p>

<pre>CODE</pre>

<p>Eso es un ejemplo de una vulnerabilidad grave en la seguridad de una aplicación (web o no) porque si el usuario introdujese un valor como CODE nos encontraríamos con que la sentencia ejecutada sería:</p>

<pre>CODE</pre>

<p>Y se eliminaría la tabla Usuarios con todos los datos contenidos en ella. </p>

<p><strong>¿Cómo puedo evitar que la inyección SQL ocurra en PHP?</strong></p>



In [13]:
questoes_pt_sem_codigo = substitui(regex_cod, dados_pt.Questão, "CODE")

In [14]:
questoes_pt_sem_cod_tag = remover(regex_html, questoes_pt_sem_codigo)

In [15]:
dados_pt["sem_code_tag"] = questoes_pt_sem_cod_tag

In [16]:
dados_pt.head()

Id  ...                                       sem_code_tag
0   2402  ...  Se eu fizer o hash de senhas antes de armazená...
1   6441  ...  Qual é a diferença entre CODE e CODE? Podem me...
2    579  ...  Uma dúvida muito comum é por que devemos parar...
3   2539  ...  É comum encontrar uma mensagem de erro que diz...
4  17501  ...  Me parecem termos muito próximos e eventualmen...

[5 rows x 7 columns]

Tirando das outras questões.

In [17]:
questoes_ing_sem_codigo = substitui(regex_cod, dados_ing.Questão, "CODE")
questoes_ing_sem_cod_tag = remover(regex_html, questoes_ing_sem_codigo)
dados_pt["sem_code_tag"] = questoes_ing_sem_cod_tag

questoes_sp_sem_codigo = substitui(regex_cod, dados_sp.Questão, "CODE")
questoes_sp_sem_cod_tag = remover(regex_html, questoes_sp_sem_codigo)
dados_pt["sem_code_tag"] = questoes_sp_sem_cod_tag

Remover a pontuação, pois para o projeto de identificação da linguagem deveria levar-se em conta apenas as palavras.

In [18]:
# Entre [] e com ^ tira tudo menos aquilo que for escrito, no caso do \w tira tudo menos os caracteres alfa-numéricos
# é preciso que coloque o \s para que não seja removido junto com isso os espaços e quebras de linha.
regex_pontos = re.compile(r"[^\w\s]")
questoes_pt_sem_codtagpont = remover(regex_pontos, questoes_pt_sem_cod_tag)
print(questoes_pt_sem_codtagpont[0])

Se eu fizer o hash de senhas antes de armazenálas em meu banco de dados é suficiente para evitar que elas sejam recuperadas por alguém

Estou falando apenas da recuperação diretamente do banco de dados e não qualquer outro tipo de ataque como força bruta na página de login da aplicação keylogger no cliente e criptoanálise rubberhose Qualquer forma de hash não vai impedir esses ataques

Tenho preocupação em dificultar ou até impossibilitar a obtenção das senhas originais caso o banco de dados seja comprometido Como dar maior garantia de segurança neste aspecto

Quais preocupações adicionais evitariam o acesso às senhas Existem formas melhores de fazer esse hash



In [19]:
# Como não possui diferença entre carcteres minusculos e maiusculos é interessante transformarmos tudo em minusculo
# para que no processador compreenda que são palavras iguais.
def minusculo (textos):
  if type(textos) == str:
    return textos.lower()
  else:
    return [texto.lower() for texto in textos]
questoes_pt_codtagpontmin = minusculo(questoes_pt_sem_codtagpont)
print(questoes_pt_codtagpontmin[0])

se eu fizer o hash de senhas antes de armazenálas em meu banco de dados é suficiente para evitar que elas sejam recuperadas por alguém

estou falando apenas da recuperação diretamente do banco de dados e não qualquer outro tipo de ataque como força bruta na página de login da aplicação keylogger no cliente e criptoanálise rubberhose qualquer forma de hash não vai impedir esses ataques

tenho preocupação em dificultar ou até impossibilitar a obtenção das senhas originais caso o banco de dados seja comprometido como dar maior garantia de segurança neste aspecto

quais preocupações adicionais evitariam o acesso às senhas existem formas melhores de fazer esse hash



In [20]:
# Para remover os digitos usamos \d e o + para significar que pode ser presente 1 ou mais vezes
regex_digitos = re.compile(r"\d+")
questoes_pt_codtagpontmindig = remover(regex_digitos, questoes_pt_codtagpontmin)
print(questoes_pt_codtagpontmindig[0])

se eu fizer o hash de senhas antes de armazenálas em meu banco de dados é suficiente para evitar que elas sejam recuperadas por alguém

estou falando apenas da recuperação diretamente do banco de dados e não qualquer outro tipo de ataque como força bruta na página de login da aplicação keylogger no cliente e criptoanálise rubberhose qualquer forma de hash não vai impedir esses ataques

tenho preocupação em dificultar ou até impossibilitar a obtenção das senhas originais caso o banco de dados seja comprometido como dar maior garantia de segurança neste aspecto

quais preocupações adicionais evitariam o acesso às senhas existem formas melhores de fazer esse hash



In [21]:
# Com as regex aplicadas podem ocorrer casos de espaços duplos, então é interessante fazer uma regex para esses casos
regex_espaco = re.compile(r" +")
exemplo = ("Texto    de      teste")
exemplo_1 = substitui(regex_espaco, exemplo, " ")
print(exemplo_1)

Texto de teste


In [22]:
# Aplicando nas questões pt
questoes_pt_atualizada = substitui(regex_espaco, questoes_pt_codtagpontmindig, " ")
print(questoes_pt_atualizada[0])

se eu fizer o hash de senhas antes de armazenálas em meu banco de dados é suficiente para evitar que elas sejam recuperadas por alguém

estou falando apenas da recuperação diretamente do banco de dados e não qualquer outro tipo de ataque como força bruta na página de login da aplicação keylogger no cliente e criptoanálise rubberhose qualquer forma de hash não vai impedir esses ataques

tenho preocupação em dificultar ou até impossibilitar a obtenção das senhas originais caso o banco de dados seja comprometido como dar maior garantia de segurança neste aspecto

quais preocupações adicionais evitariam o acesso às senhas existem formas melhores de fazer esse hash



In [23]:
# A mesma lógica dos espaços pode ser usada para o tratamento quebras de linha, visto que não são necessárias.
regex_quebralinha = re.compile(r"(\n)+")
questoes_pt_atualizada = substitui(regex_quebralinha, questoes_pt_atualizada, " ")
print(questoes_pt_atualizada[0])

se eu fizer o hash de senhas antes de armazenálas em meu banco de dados é suficiente para evitar que elas sejam recuperadas por alguém estou falando apenas da recuperação diretamente do banco de dados e não qualquer outro tipo de ataque como força bruta na página de login da aplicação keylogger no cliente e criptoanálise rubberhose qualquer forma de hash não vai impedir esses ataques tenho preocupação em dificultar ou até impossibilitar a obtenção das senhas originais caso o banco de dados seja comprometido como dar maior garantia de segurança neste aspecto quais preocupações adicionais evitariam o acesso às senhas existem formas melhores de fazer esse hash 


In [24]:
# Tendo feito todas as limpezas possíveis vou fazer o mesmo para as outras linguas e colocar uma nova coluna no dataset
dados_pt["questao_tratada"] = questoes_pt_atualizada

questoes_ing_atualizada = questoes_ing_sem_cod_tag
questoes_ing_atualizada = remover(regex_pontos, questoes_ing_atualizada)
questoes_ing_atualizada = minusculo(questoes_ing_atualizada)
questoes_ing_atualizada = remover(regex_digitos, questoes_ing_atualizada)
questoes_ing_atualizada = substitui(regex_quebralinha, questoes_ing_atualizada, " ")
questoes_ing_atualizada = substitui(regex_espaco, questoes_ing_atualizada, " ")
dados_ing["questao_tratada"] = questoes_ing_atualizada

questoes_sp_atualizada = questoes_sp_sem_cod_tag
questoes_sp_atualizada = remover(regex_pontos, questoes_sp_atualizada)
questoes_sp_atualizada = minusculo(questoes_sp_atualizada)
questoes_sp_atualizada = remover(regex_digitos, questoes_sp_atualizada)
questoes_sp_atualizada = substitui(regex_quebralinha, questoes_sp_atualizada, " ")
questoes_sp_atualizada = substitui(regex_espaco, questoes_sp_atualizada, " ")
dados_sp["questao_tratada"] = questoes_sp_atualizada

In [25]:
dados_ing.head()

Id  ...                                    questao_tratada
0  11227809  ...  here is a piece of c code that seems very pecu...
1    927358  ...  i accidentally committed the wrong files to gi...
2   2003505  ...  i want to delete a branch both locally and rem...
3    292357  ...   moderator note given that this question has a...
4    477816  ...  ive been messing around with json for some tim...

[5 rows x 7 columns]

#Modelo de Linguagem

O modelo de linguagem é uma tentativa de prever o que virá a seguir baseado numa informação prévia, ou seja baseado naquilo que já foi visto ele calculará a probabilidade e "advinhará" o próximo evento que irá acontecer. E para o nosso projeto, baseado no resultado apresentado ele irá identificar em qual idioma o texto foi escrito.

In [52]:
from sklearn.model_selection import train_test_split
from nltk.tokenize import WhitespaceTokenizer
from nltk.lm.preprocessing import padded_everygram_pipeline
from nltk.lm import MLE
from nltk.util import bigrams
from nltk.lm.preprocessing import pad_both_ends
from nltk.lm import Laplace

In [27]:
# Para depois identificarmos de qual idioma é cada questão, vamos criar uma coluna indicando isso.
dados_pt["idioma"] = "port"
dados_ing["idioma"] = "ing"
dados_sp["idioma"] = "esp"

In [28]:
# Separando em treino e teste
port_treino, port_teste = train_test_split(dados_pt.questao_tratada, test_size = 0.2, random_state = 123)
ing_treino, ing_teste = train_test_split(dados_ing.questao_tratada, test_size = 0.2, random_state = 123)
sp_treino, sp_teste = train_test_split(dados_sp.questao_tratada, test_size = 0.2, random_state = 123)

In [29]:
# Para começar o modelo faremos uma variavel com todas as questões separadas apenas por espaço para a analise por parte
# do modelo. E posteriormente tokenizar para cada palavra ser um item de uma lista.

todas_as_questoes_port = ' '.join(port_treino)
todas_as_palavras_port = WhitespaceTokenizer().tokenize(todas_as_questoes_port)

In [30]:
# Iremos separar as palavras em bigramas e colocar um fakechar no inicio e fim de cada palavra para evitar que dentro
# do bigrama a letra inicial e final da palavra fique apenas em um bigrama e não em dois como as outras criando um viés
# no modelo

port_treino_bigram, vocab_port = padded_everygram_pipeline(2, todas_as_palavras_port)

In [31]:
# Criando o modelo e treinando
model_port = MLE(2)
model_port.fit(port_treino_bigram, vocab_port)

##Treinamento do modelo

In [33]:
def treinar_modelo_mle(lista_textos):
  frases_juntas = ' '.join(lista_textos)
  palavras_juntas = WhitespaceTokenizer().tokenize(frases_juntas)
  texto_bigram, vocab_texto = padded_everygram_pipeline(2, palavras_juntas)
  model = MLE(2)
  model.fit(texto_bigram, vocab_texto)
  return model

In [34]:
model_port = treinar_modelo_mle(port_treino)
model_ing = treinar_modelo_mle(ing_treino)
model_sp = treinar_modelo_mle(sp_treino)

##Construindo função para calcular a perplexidade de uma frase

Perplexidade é o nome da métrica capaz de medir a proximidade que os bigramas de uma palavra tem em relação aos bigramas usados para treinar o modelo, no caso, quanto menor a perplexidade mais semelhante com o treinado.

In [49]:
def calcula_perplexidade(model, texto):
  perplexidade = 0
  palavras = WhitespaceTokenizer().tokenize(texto)
  palavras_fakechar = [list(pad_both_ends(palavra, n=2)) for palavra in palavras]
  palavras_bigrams = [list(bigrams(palavra)) for palavra in palavras_fakechar]

  for palavra in palavras_bigrams:
    perplexidade += model.perplexity(palavra)
  
  return perplexidade

In [50]:
print(calcula_perplexidade(model_ing, "good morning"))
print(calcula_perplexidade(model_sp, "good morning"))
print(calcula_perplexidade(model_port, "good morning"))

30.1620889594439
54.55542265319791
69.20759508807609


Conforme podemos observar, atraves das perplexidades calculadas com nossos modelos, que a menor perplexidade foi a do modelo ing, logo chegamos a conclusão que a frase "good morning" pertence a lingua inglesa.

In [51]:
# Testeando para uma questão da nossa base de dados
print(calcula_perplexidade(model_ing, port_teste.iloc[0]))
print(calcula_perplexidade(model_sp, port_teste.iloc[0]))
print(calcula_perplexidade(model_port, port_teste.iloc[0]))

inf
inf
2006.9786364086417


Podemos notar que nos modelos em espanhol e inglês a perplexidade deu infinito, isso se dá porque o calculo da perplexidade é o inverso da probabilidadem ou seja como alguma palavra ou bigrama da frase não foi localizado nenhuma vez nos modelos de outra lingua a perplexidade ficou como 1/0 que levou o total a infinito.

##Implementação Laplace

Tecnica de Laplace: é um modelo que serve para impedir que a probabilidade seja zeroe assim impede que a perplexidade seja infinita.

In [53]:
def treinar_modelo_Laplace(lista_textos):
  frases_juntas = ' '.join(lista_textos)
  palavras_juntas = WhitespaceTokenizer().tokenize(frases_juntas)
  texto_bigram, vocab_texto = padded_everygram_pipeline(2, palavras_juntas)
  model = Laplace(2)
  model.fit(texto_bigram, vocab_texto)
  return model

In [55]:
# Treinar novos modelos com o Laplave e realizar o mesmo teste com a questão em português
model_port2 = treinar_modelo_Laplace(port_treino)
model_ing2 = treinar_modelo_Laplace(ing_treino)
model_sp2 = treinar_modelo_Laplace(sp_treino)

print(calcula_perplexidade(model_ing2, port_teste.iloc[0]))
print(calcula_perplexidade(model_sp2, port_teste.iloc[0]))
print(calcula_perplexidade(model_port2, port_teste.iloc[0]))

5876.837588345698
3488.5698949157722
2009.1937946178912


##Aplicação do modelo nos dados de teste.

In [56]:
def atribui_idioma(lista_textos):
  idioma = []
  for texto in lista_textos:
    ingles = calcula_perplexidade(model_ing2, texto)
    espanhol = calcula_perplexidade(model_sp2, texto)
    portugues = calcula_perplexidade(model_port2, texto)
    if ingles >= portugues <= espanhol:
      idioma.append("portugues")
    elif portugues >= ingles <= espanhol:
      idioma.append("ingles")
    else: 
      idioma.append("espanhol")
  return idioma

In [60]:
resultados_port = atribui_idioma(port_teste)
print(resultados_port.count("portugues"))

100


Tivemos uma acurácia de 100% na identificação do idioma das questões em português.

In [62]:
resultados_ing = atribui_idioma(ing_teste)
print(resultados_ing.count("ingles"))

print("***************")

resultados_sp = atribui_idioma(sp_teste)
print(resultados_sp.count("espanhol"))
print(resultados_sp.count("portugues"))
print(resultados_sp.count("ingles"))

100
***************
97
1
2


Para as questões em inglês tivemos 100% de acerto também, e em espanhol tivemos 97% de acerto com 1 questão identificada equivocadamente como português e 2 para o inglês